eth/downloader/whitelist/service.go

```diff
@@ -70,6 +72,96 @@ func (w *Service) IsValidChain(remoteHeader *types.Header, fetchHeadersByNumber
 	return false, ErrCheckpointMismatch
 }
 
+// IsValidChain checks the validity of chain by comparing it
+// against the local checkpoint entries
+func (w *Service) IsValidChain(currentHeader *types.Header, chain []*types.Header) bool {
+	var (
+		oldestCheckpointNumber uint64 = w.checkpointOrder[0]
+		first                  uint64 = chain[0].Number.Uint64()
+		last                   uint64 = chain[len(chain)-1].Number.Uint64()
+		current                uint64 = currentHeader.Number.Uint64()
+	)
+
+	// Check if we have checkpoints to validate incoming chain in memory
+	if len(w.checkpointWhitelist) == 0 {
+		// We don't have any entries, no additional validation will be possible
+		return true
+	}
+
+	// Check if we have whitelist entries in required range
+	if last < oldestCheckpointNumber {
+		// We have future whitelisted entries, so no additional validation will be possible
+		// This case will occur when bor is in middle of sync, but heimdall is ahead/fully synced.
+		return true
+	}
+
+	// Start iterating over the chain and apply validation for past and future
+	// chain, referring to chain behind and after current header respectively.
+	for i := 0; i < len(chain); i++ {
+	}
+
+	var (
+		pastChain   []*types.Header
+		futureChain []*types.Header
+	)
+	if current >= first {
+		if len(chain) == 1 {
+			pastChain = chain
+		} else {
+			pastChain = chain[:current-first]
+		}
+	}
+	if current < last {
+		if len(chain) == 1 {
+			futureChain = chain
+		} else {
+			futureChain = chain[current-first+1:]
+		}
+	}
+
+	if len(futureChain) > int(w.checkpointInterval) {
+		return false
+	}
+
+	// Iterate over the chain and validate against the last checkpoint
+	// It will handle all cases where the incoming chain has atleast one checkpoint
+	for i := len(pastChain) - 1; i >= 0; i-- {
+		if _, ok := w.checkpointWhitelist[pastChain[i].Number.Uint64()]; ok {
+			if pastChain[i].Hash() != w.checkpointWhitelist[pastChain[i].Number.Uint64()] {
+				return false
+			} else {
```

Comment:
you don't need `else` here as far as if-statement ends with `return`, just `break` if enough.

Reply:
You mean something like this?
`return pastChain[i].Hash() == w.checkpointWhitelist[pastChain[i].Number.Uint64()]`?